internal/backend/usagereporter/usagedata.go

```diff
@@ -0,0 +1,145 @@
+package usagereporter
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/google/uuid"
+	"go.autokitteh.dev/autokitteh/internal/version"
+	"go.autokitteh.dev/autokitteh/internal/xdg"
+	"go.uber.org/zap"
+)
+
+type reportRequest struct {
+	InstallationID string `json:"installation_id"`
+	Version        string `json:"version"`
+	Commit         string `json:"commit"`
+	OS             string `json:"os"`
+	Arch           string `json:"arch"`
+}
+
+type UsageReporter interface {
+	Start()
+	Stop()
+}
+
+type usageReporter struct {
+	installationID uuid.UUID
+	updateInterval time.Duration
+	shutdownChan   chan struct{}
+	endpoint       string
+	logger         *zap.Logger
+}
+
+func generateInstallIDFile(path string) error {
+	id := uuid.New()
+	if err := os.WriteFile(path, []byte(id.String()), 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+func ensureAndReadInstallationIDFile(installationIDFile string) (string, error) {
+	// Try read installationID file,
+	// if there is an error and the file exists, try to remove it and fail if cant
+	// generate a new file
+	// try read the data again
+	data, err := os.ReadFile(installationIDFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			if err := os.Remove(installationIDFile); err != nil {
+				return "", err
+			}
+		}
+
+		if err := generateInstallIDFile(installationIDFile); err != nil {
+			return "", err
+		}
+		if data, err = os.ReadFile(installationIDFile); err != nil {
+			return "", err
+		}
+
+	}
+
+	return string(data), nil
+}
+
+func New(z *zap.Logger, cfg *Config) (UsageReporter, error) {
+	if !cfg.Enabled {
+		return &nopUpdater{}, nil
+	}
+
+	installationIDFile := filepath.Join(xdg.ConfigHomeDir(), "installID")
```

Comment:
I'm not sure ConfigHomeDir is the right choice - maybe DataHomeDir because it's not exactly "configurable"?

But it's up to you.

Reply:
changed